### PR TITLE
WanOperation for delivery control in enterprise wan replication

### PR DESCRIPTION
**Pull request by emrahkocaman**
Also changes `OperationRunnerImpl` and `OperationAccessor` to allow WanOperation
to run on non-member targets

2nd part of https://github.com/hazelcast/hazelcast/pull/5079

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java [hunk 1]
```diff
@@ -311,7 +312,9 @@ public void run(Packet packet) throws Exception {
     }
 
     private boolean ensureValidMember(Operation op) {
-        if (isJoinOperation(op) || node.clusterService.getMember(op.getCallerAddress()) != null) {
+        if (isJoinOperation(op)
+                || isWanReplicationOperation(op)
+                || node.clusterService.getMember(op.getCallerAddress()) != null) {
```
Comment (jerrinot) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java]: this can be short-circuited as well - as majority of operations have a valid caller set.

**Comment by emrahkocaman** [comment]
verify

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java (hunk 1 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java]: Done.

File: hazelcast/src/main/java/com/hazelcast/spi/OperationAccessor.java [hunk 2]
```diff
@@ -43,6 +44,11 @@ public static boolean isMigrationOperation(Operation op) {
                 && op.getClass().getClassLoader() == THIS_CLASS_LOADER;
     }
 
+    public static boolean isWanReplicationOperation(Operation op) {
+        return op instanceof WanReplicationOperation
+                && op.getClass().getClassLoader() == THIS_CLASS_LOADER;
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/OperationAccessor.java]: Why do you add this to the OperationAccessor? And why does it matter which classloaded loaded it?

Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/OperationAccessor.java]: Ok. I see that already is being done on the OperationAccessor by other functions. Not very nice. Can you move all of them to a utility class called e..g Operations?

Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/spi/OperationAccessor.java]: Ok. Moved to Operations.

File: hazelcast/src/main/java/com/hazelcast/spi/Operations.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+        return op instanceof JoinOperation
+                && op.getClass().getClassLoader() == THIS_CLASS_LOADER;
+    }
+
+    public static boolean isMigrationOperation(Operation op) {
+        return op instanceof MigrationCycleOperation
+                && op.getClass().getClassLoader() == THIS_CLASS_LOADER;
+    }
+
+    public static boolean isWanReplicationOperation(Operation op) {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operations.java]: Can you add Javadoc.

Can you ask @mdogan why the classloader is added to deal with marker interfaces? Because we are using marker interfaces at other places without class loader check.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java [hunk 4]
```diff
@@ -239,7 +240,7 @@ protected void ensureNoPartitionProblems(Operation op) {
     }
 
     private boolean retryDuringMigration(Operation op) {
-        return !(op instanceof ReadonlyOperation || OperationAccessor.isMigrationOperation(op));
+        return !(op instanceof ReadonlyOperation || Operations.isMigrationOperation(op));
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java]: Can you use a static import. So the Operations. stuff isn't needed anymore. I'm trying to apply them where possible to make code easier to read (less long lines).

File: hazelcast/src/main/java/com/hazelcast/spi/Operations.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+package com.hazelcast.spi;
+
+import com.hazelcast.cluster.impl.operations.JoinOperation;
+import com.hazelcast.cluster.impl.operations.WanReplicationOperation;
+import com.hazelcast.partition.MigrationCycleOperation;
+
+/**
+ * Utility class that contains helper methods related to {@link Operation}
+ */
+public final class Operations {
```
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/spi/Operations.java]: Can you move this whole class into the com.hazelcast.spi.impl.operations package? Afaik user doesn't need to see this and we need to prevent exposing new API's if not needed

File: hazelcast/src/main/java/com/hazelcast/spi/Operations.java (hunk 3 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/spi/Operations.java]: Javadoc added. I don't think we can add tests for `isWanReplicationOperation` in OSS as there is no concrete classes implementing `WanReplicationOperation`.

File: hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java (hunk 4 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/spi/impl/operationservice/impl/OperationRunnerImpl.java]: Done, even this is out of scope of this PR.

File: hazelcast/src/main/java/com/hazelcast/spi/Operations.java (hunk 5 above)
Comment (emrahkocaman) [inline hazelcast/src/main/java/com/hazelcast/spi/Operations.java]: Moved to "com.hazelcast.spi.impl.operationutil" package.

**Comment by mesutcelik** [comment]
verify

**Comment by mdogan** [comment]
LGTM :+1:
